istioctl/cmd/istiodconfig.go

```diff
@@ -0,0 +1,464 @@
+	}
+	return scopeInfos, nil
+}
+
+func (c *ControlzClient) PutScope(scope *ScopeInfo) error {
+	var jsonScopeInfo bytes.Buffer
+	err := json.NewEncoder(&jsonScopeInfo).Encode(scope)
+	if err != nil {
+		return fmt.Errorf("cannot serialize scope %+v", *scope)
+	}
+	req, err := http.NewRequest(http.MethodPut, c.baseURL.String()+"/"+scope.Name, &jsonScopeInfo)
+	if err != nil {
+		return err
+	}
+	defer req.Body.Close()
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusAccepted {
+		return fmt.Errorf("cannot update resource %s, got status %s", scope.Name, resp.Status)
+	}
+	return nil
+}
+
+func (c *ControlzClient) GetScope(scope string) (*ScopeInfo, error) {
+	var s ScopeInfo
+	resp, err := http.Get(c.baseURL.String() + "/" + scope)
+	if err != nil {
+		return &s, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return &s, fmt.Errorf("request not successful %s: ", resp.Status)
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&s)
+	if err != nil {
+		return &s, fmt.Errorf("cannot deserialize response: %s", err)
+	}
+	return &s, nil
+}
+
+var (
+	istiodLabelSelector = ""
+	istiodReset         = false
+)
+
+func istiodLogCmd() *cobra.Command {
+	var opts clioptions.ControlPlaneOptions
+	var outputLogLevel = ""
+	var stackTraceLevel = ""
+
+	// output format (yaml or short)
+	var outputFormat = "short"
+
+	logCmd := &cobra.Command{
+		Use:   "log [--level <scope>:<level>][--stack-trace-level <scope>:<level>]|[-r|--reset]|[--output|-o short(default)|yaml]",
+		Short: "Manage istiod logging.",
+		Long:  "Retrieve or update logging levels of istiod components.",
+		Example: ` # Retrieve information about istiod logging levels.
+  istioctl istiod-config log [<pod-name>]
+
+  # Update levels of the specified loggers.
+  istioctl istiod-config log --level ads:debug,authorization:debug
+
+  # Reset levels of all the loggers to default value (info).
+  istioctl istiod-config log -r
+`,
+		Aliases: []string{"o"},
+		Args: func(logCmd *cobra.Command, args []string) error {
+			if istiodReset && outputLogLevel != "" {
+				logCmd.Println(logCmd.UsageString())
+				return fmt.Errorf("--level cannot be combined with --reset")
+			}
+			if istiodReset && stackTraceLevel != "" {
+				logCmd.Println(logCmd.UsageString())
+				return fmt.Errorf("--stack-trace-level cannot be combined with --reset")
+			}
+			return nil
+		},
+		RunE: func(logCmd *cobra.Command, args []string) error {
+			client, err := kubeClientWithRevision(kubeconfig, configContext, opts.Revision)
+			if err != nil {
+				return fmt.Errorf("failed to create k8s client: %v", err)
+			}
+
+			var podName, ns string
+			if len(args) == 0 {
+				pl, err := client.PodsForSelector(context.TODO(), handlers.HandleNamespace(istioNamespace, defaultNamespace), istiodLabelSelector)
+				if err != nil {
+					return fmt.Errorf("not able to locate pod with selector %s: %v", istiodLabelSelector, err)
```

Comment:
help text says `istioctl istiod-config log [<pod-name>]`, but code says label selector. IMO, it should be either (or both):
* Take pod name, set log level on that one. Prefer this to match the proxy-config command, you may want to debug single instance
* Take label, set log level on all of them

IMO we should not use label selector then only use one of them

Reply:
Square brackets around pod-name suppose to suggest it's optional parameter. Currently user can provide podname or selector or neither (default selector app=istiod will be used). Currently if there are multiple pods with label app=istiod only first one will used as in:
https://github.com/istio/istio/blob/d89c6a624cfd0962e2ed25238ec671baebda0ca5/istioctl/cmd/dashboard.go#L294-L296
I will improve the first line (Usage) of help for log subcommand

Reply:
Square brackets around pod-name suppose to suggest it's optional parameter. Currently user can provide podname or selector or neither (default selector app=istiod will be used). Currently if there are multiple pods with label app=istiod only first one will used as in:

istio/istioctl/cmd/dashboard.go

Lines 294 to 296 in d89c6a6
 if len(pl.Items) > 1 { 
 	log.Warnf("more than 1 pods fits selector: %s; will use pod: %s", labelSelector, pl.Items[0].Name) 
 } 

I will improve the first line (Usage) of help for log subcommand